New user rest APIs

polothy opened the pull request.

kaulavinash committed:
Additional ILP API to get grades and student activity from Moodle

kaulavinash committed:
Final changes for review after refactoring comments and removing unused code.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+
+		if (strlen($studentIds) > 0) {
+			//add quotes to incoming list
+			$userids = explode(",", $studentIds);
```
Must clean user IDs.  Use:

```
$userids = clean_param_array($userids, PARAM_INT);
```

avinashkellucian replied:
The query uses idnumber attribute to query the user's attributes. The idnumber can be alphanumeric (any string value) in Moodle.

polothy replied:
Then this should work:

```
$userids = clean_param_array($userids, PARAM_NOTAGS);
```

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+		}
+	
+
+		foreach ($userids as $userid) {
```
Performance of this is troubling.  How many users might you be passing to this method?  We might hit around 10+ queries per user per enrolled course and not all queries are light.

Anyway we can restrict this to a single user?

avinashkellucian replied:
Even though we are passing an list of student/user ids to the API. We query the data one user at a time. We need this data for all students on from the ERP. We do batch the user request and can tuned it based on our performance tests.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+	
+
+		foreach ($userids as $userid) {
+			$sections = self :: get_courses_where_user_is_enrolled($userid , $startDate);
```
This method is not entirely accurate.  For example, a user may not be able to see the course because it is hidden or the category is hidden.  I recommend that you use $this->helper->connector->get_courses(...) (That goes to the block_intelligent_learning_helper_connector class).  Or you can just use enrol_get_users_courses(...) directly.

You can then just skip courses that do not meet your date restriction.

avinashkellucian replied:
I moved this method to the helper->connector.php. I created a new helper method get_courses_where_user_is_enrolled($userid, $startdate = null) in the connector.php.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+			                  FROM {user} u
+			                  WHERE u.idnumber = :userid";
+
+			$user = $DB->get_record_sql($sql, $sql_param);
```
Move this outside of the "foreach ($sections as $section){" loop as it will be the same for each course.

avinashkellucian replied:
Refactored and corrected the code.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+	* 
+	* @return array
+	*/
+	public function aboutMoodle() {
```
Currently we are not using camel case for methods, recommend renaming to about_moodle.

avinashkellucian replied:
Fixed

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+	*/
+	public function aboutMoodle() {
+		
+		$result['Version'] = $CFG->version;
```
Very strongly recommend using moodle_major_version() instead of $CFG->version as this can expose the exact version of Moodle and then attackers can use known security issues to attack the site.

avinashkellucian replied:
Fixed

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+			$select .= " AND (time < :enddate)";
+		}
+
+		$result = $DB->count_records_select('log', $select, $sql_param);
```
Log table is notoriously slow on client sites due to its size.  Could you perhaps use the event table instead?

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+		                        FROM {quiz} q
+		                        JOIN {quiz_attempts} qa ON q.id = qa.quiz
+		                	   WHERE q.course $course_sql
+		                         AND qa.userid = :userid";
```
Do you want to add a WHERE qa.state = quiz_attempt::IN_PROGRESS?  Otherwise, this will count all open, closed, overdue and abandoned attempts.  Just unsure what exactly must be counted.

avinashkellucian replied:
Didn't realize there was a state maintained for quiz. I have updated the query to check for the  'in progress' state.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+			$sql .= " AND l.time < :enddate";
+		}
+
+		return $DB->count_records_sql($sql, $sql_param);
```
Just some notes, in 2.7 the log table is being deprecated and this will likely be a slow query.  There are no great alternatives.  One would be to use completion tracking and set view as one of the criteria, but that would require every assignment to have that.  Just wondering the importance of this count as the maintenance of it after 2.7 could become tricky.  Easier to never add it than to remove it down the road.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+			$sql .= " AND l.time < :enddate";
+		}
+
+		return $DB->count_records_sql($sql, $sql_param);
```
Same comment as above.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+			$sql .= " AND l.time < :enddate";
+		}
+
+		return $DB->count_records_sql($sql, $sql_param);
```
Same comment as above about log table going out of style.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+			$sql_params['enddate'] = $enddate +DAYSECS;
+			$where .= " AND time < :enddate";
+		}
+		return $DB->count_records_select('log', $where, $sql_params);
```
Same comment as above about log table going out of style.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+	}
+	
+	
+	private static function get_student_roleid() {
```
Remove this method once get_courses_where_user_is_enrolled is removed.  It is inaccurate as well since sites can have multiple student roles.

avinashkellucian replied:
Fixed. Moved the  get_courses_where_user_is_enrolled to the helper connector php.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+	* @return integer
+	*/
+
+	private static function get_courseid_from_idnumber($sectionsourcedIds) {
```
This is not used, please remove.

avinashkellucian replied:
Fixed

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+					$UserOutput .= ",";
+				}
+
+				$UserOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
SQL injection problem here.  Must use PARAM_INT and add the variable to the $params so the data binding is done in the DB layer.

avinashkellucian replied:
The userids and sectionsids can be of text values.
Should I clean the array using the clean_param_array($userids, PARAM_NOTAGS) API ?

polothy replied:
Yeah, sorry, I thought again it was user.id and not user.idnumber.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+		// Add Userids to the query
+		if (strlen($userIds) > 0) {
+			//add quotes to incoming list
+			$UserIDList = explode(",", $userIds);
```
You can clean all user IDs with clean_param_array

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+					$SectionOutput .= ",";
+				}
+
+				$SectionOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
Same SQL injection problem here

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+		if (count($updateData) > 0) {
+			$sectionIds = implode(",", $updateData);
+			$newData = blocks_intelligent_learning_model_service_user :: get_user_grades($userIds, $sectionIds, $startDate, $endDate, $RecursiveLevel++);
+			$returnData = array_merge($newData, $returnData);
```
I'm not sure if this makes sense because get_user_grades will return the formatted response, not the raw data.

Also, should use $this-> instead of blocks_intelligent_learning_model_service_user:: since it is not a static method.

avinashkellucian replied:
What the difference between the formatted response and raw data. Are you referring to just the format of the response?
That was done intentionally. We may have to circle back on this. 

Fixed the incorrect static call.

polothy replied:
The get_user_grades method returns the DOMDocument, not an array of data.  So, you cannot merge the two together.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+					$UserOutput .= ",";
+				}
+
+				$UserOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
SQL injection problem here

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+					$SectionOutput .= ",";
+				}
+
+				$SectionOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
SQL injection problem here

polothy commented on model/service/user.php:
```diff
@@ -479,4 +479,908 @@ private function update($user, $data = NULL) {
+					$SectionOutput .= ",";
+				}
+
+				$SectionOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
SQL injection problem here

kaulavinash committed:
Fixed Review comments from Moodlerooms.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +480,824 @@ private function update($user, $data = NULL) {
+                    $UserOutput .= ",";
+                }
+
+                $UserOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
The way to convert an array of options for a DB query like this (note the below code replaces the foreach loop here):

```
$UserIDList = explode(",", $userIds);
$UserIDList = clean_param_array($UserIDList, PARAM_NOTAGS);
list($extrasql, $extraparams) = $DB->get_in_or_equal($UserIDList);
$sql .= $addAnd . ' {course}.idnumber ' . $extrasql;
$params = array_merge($params, $extraparams);
```

Not doing this opens up SQL injection problems.  Also, if $id is the idnumber, then please use PARAM_NOTAGS instead of PARAM_TEXT.

polothy commented on model/service/user.php:
```diff
@@ -479,4 +480,824 @@ private function update($user, $data = NULL) {
+                    $SectionOutput .= ",";
+                }
+
+                $SectionOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
Please see above comment regarding $DB->get_in_or_equal

polothy commented on model/service/user.php:
```diff
@@ -479,4 +480,824 @@ private function update($user, $data = NULL) {
+                    $UserOutput .= ",";
+                }
+
+                $UserOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
Please see above comment regarding $DB->get_in_or_equal

polothy commented on model/service/user.php:
```diff
@@ -479,4 +480,824 @@ private function update($user, $data = NULL) {
+                    $SectionOutput .= ",";
+                }
+
+                $SectionOutput .= "'" . clean_param($id, PARAM_TEXT) . "'";
```
Please see above comment regarding $DB->get_in_or_equal

ellucianmoodle committed:
Test commit

Test commit into 4.2 branch.

ellucianmoodle committed:
ILP 4.2 - Real-time grades posting

Changes to support real-time grades posting.

ellucianmoodle committed:
ILP 4.2 bug fixes

Fix formatting and remove tabs; course service bug fixes

ellucianmoodle committed:
Code review fixes

Code review fixes; updates to empty/null values for dates

ellucianmoodle committed:
Error handling improvements

ellucianmoodle committed:
Bug fixes

ellucianmoodle committed:
Fix display issue with dates

kaulavinash committed:
Defect fixes

kaulavinash committed:
Formatted - removed tab spaces

kaulavinash committed:
More review comment fixes. Changed SQL injection

ellucianmoodle committed:
Final check-in for ILP 4.2

ellucianmoodle committed:
Merge conflict

ellucianmoodle committed:
Remove merge tag

kaulavinash committed:
Merge fixes

Added missing lines and changed to reuse ILP code

dsavkoor committed:
Update version to match that in the last build. All else is in sync.

dsavkoor committed:
Modifications to implement gradebook changes from Moodle v3.0 up

ellucianmoodle committed:
Merge pull request #1 from EllucianMoodleIntegration/ILP-Moodle-3.0.0

Modifications to implement gradebook changes from Moodle v3.0 up

avinashkellucian committed:
Updates to get assignment submitted query

This updates resolve the incorrect count returned for number of student
assignments submitted

dsavkoor committed:
Remove numsection references and fixes for concurrent section/enrolment creation issues

dsavkoor committed:
Retract Advise change from 3.0.0 branch to avoid merge conflict

ellucianmoodle committed:
Merge pull request #2 from EllucianMoodleIntegration/ILP-Moodle-3.0.0

Remove numsection references and fixes for concurrent section/enrolme…

dsavkoor committed:
Handling of course categories during ILP bulk processing and course end dates.

ellucianmoodle committed:
Merge pull request #3 from EllucianMoodleIntegration/ILP-Moodle-3.0.0

Handling of course categories during ILP bulk processing and course e…

dsavkoor committed:
Defect fixes for honoring numsections and visibility on course creation and removal of deprecated showavailability property.

dsavkoor committed:
Retract Advise change to avoid merge conflicts

dsavkoor committed:
Retract advise change- take 2

ellucianmoodle committed:
Merge pull request #4 from EllucianMoodleIntegration/ILP-Moodle-3.0.0

Defect fixes for honoring numsections and visibility on course creation

dsavkoor committed:
v3.0.3 - Allow use of standard logging to retrieve user activities

dsavkoor committed:
Merge pull request #5 from EllucianMoodleIntegration/dev

v3.0.3 - Allow use of standard logging to retrieve user activities

dsavkoor committed:
Moodle GDPR support

dsavkoor committed:
Merge pull request #6 from EllucianMoodleIntegration/dev

Moodle GDPR support

dsavkoor committed:
Moodle GDPR compatibility support and calendar event service fix

dsavkoor committed:
Merge pull request #7 from EllucianMoodleIntegration/dev

Moodle GDPR compatibility support and calendar event service fix

dsavkoor committed:
ILP-6275: Remove deprecated function references during sso login

dsavkoor committed:
Merge pull request #8 from EllucianMoodleIntegration/dev

ILP-6275: Remove deprecated function references during sso login

ilp-dev committed:
Plugin release v3.2.3

ellucianmoodle committed:
Merge pull request #9 from nitink92/v3.2.3

Plugin release v3.2.3

ilp-dev committed:
plugin release v3.2.4

ilp-dev committed:
plugin release v3.2.5

ilp-dev committed:
plugin release v3.2.6

ilp-dev committed:
plugin release v3.2.7

ilp-dev committed:
added logic fix for category-cutoff

ilp-dev committed:
plugin release v3.2.8

ellucianmoodle committed:
plugin release v3.2.9

populate currentGrade for LTI and omit duplicateBlock for courses

ellucianmoodle committed:
Updated version.php to v3.2.9

This commit is just to update the plugin version.php file to v3.2.9 and has no other changes.

ellucianmoodle committed:
Moodle plugin release 3.2.10 - Granular grading support & fix for moodle group defect
